### PR TITLE
Fix passing of links-overrides with target-applies-to-host and an implicit target

### PR DESCRIPTION
**Pull request by gmorenz**
### What does this PR try to resolve?

This fixes the link-overrides half of #14195, both the panic, and the fact that the field is being discarded, the latter of which caused the former as discussed in [the issue](https://github.com/rust-lang/cargo/issues/14195#issuecomment-2211481773).

It does so following the blueprint laid out in #13900 - which is also in my opinion the current best summary of the broader context.

### How should we test and review this PR?

For reviewing, comparing to the changes in #13900 might be useful.

### Additional information

I'm pushing a PR for the other half of #14195 simultaneously. I thought it better to keep the PRs small since they're independent, though if merged simultaneously there will be a conflict over the ordering of fields in `Unit`.

### CODE REVIEW

**Comment by gmorenz** [comment]
Well, the test doesn't work on windows, converting to draft until I fix that.

**Comment by gmorenz** [comment]
Marking this as ready for review, but leaving the fix for the test on windows as a separate comment for now because I'm hoping that there's a better way to link to a cdylib built with cargo on windows that I'm missing. See [the last commit](https://github.com/rust-lang/cargo/pull/14205/commits/0943b5c538dcb27c5e5c1b8c41ec2f93d707729e) for my current understanding of what you have to do to do that.

If there isn't a better way, I should probably squash the last commit into the first commit that introduces the test before this is merged.

File: src/cargo/core/compiler/unit.rs
Comment (weihanglo) [inline src/cargo/core/compiler/unit.rs:92]: Just realized that it should be possible to use `Rc` instead of `Arc`, for all these usages.

File: src/cargo/core/compiler/custom_build.rs
Comment (weihanglo) [inline src/cargo/core/compiler/custom_build.rs:64]: Also realized that there is an implication that these newly added fields would affect the order of any sorted units list. I don't think that would cause real issues though.

File: tests/testsuite/build_script.rs
Comment (weihanglo) [inline tests/testsuite/build_script.rs:5843]: Instead of doing this, perhaps we can go an easier route like other tests: Asserting rustflags for rustc invocation?

https://github.com/rust-lang/cargo/blob/6d685637f1853b705053500ff4abacdc51df9c91/tests/testsuite/build_script.rs#L1653-L1670

File: src/cargo/core/compiler/unit.rs
Comment (gmorenz) [reply src/cargo/core/compiler/unit.rs:92]: Huh, cargo check passes with that change at least. I think I must have just assumed they would cross a thread boundary somewhere. Will fix.

File: src/cargo/core/compiler/custom_build.rs
Comment (gmorenz) [reply src/cargo/core/compiler/custom_build.rs:64]: True... if this is a significant concern we could move the new fields to the end.

Ultimately I don't think I have the context to decide if it is. I think it is almost certainly publicly exposed given comments like [this one](https://github.com/rust-lang/cargo//blob/master/src/cargo/core/compiler/unit_dependencies.rs#L140)), and where these fields currently are in the struct is probably high enough to change the order in the cases [described in this comment](https://github.com/rust-lang/cargo//blob/master/src/cargo/core/compiler/unit.rs#L98). I wouldn't have thought that the ordering of units was something that was guaranteed to be stable between cargo versions.

I'll hold off on updating the rustflags `Arc` to `Rc` until we've decided if we need to move it to the bottom since that might as well be one PR.

File: tests/testsuite/build_script.rs
Comment (gmorenz) [reply tests/testsuite/build_script.rs:5843]: This definitely looks better, and is exactly the pre-existing test for link overrides that I failed to find. Will do this.

**Comment by gmorenz** [comment]
Made promised changes, rebased on master.

**Review by weihanglo — APPROVED** [review]
Thanks. Looks pretty great!

**Comment by weihanglo** [comment]
@bors r+

Haven't got time to review <https://github.com/rust-lang/cargo/issues/14195>, though I guess you need to rebase?

**Comment by gmorenz** [comment]
> Haven't got time to review https://github.com/rust-lang/cargo/pull/14206, though I guess you need to rebase?

Yeah, also I need to incorporate the Arc -> Rc feedback. Will get to this later today, I've just been a bit busy and was putting it off (though I don't anticipate it taking more than 10 minutes).

When you say "haven't got time" are you looking forwards and saying I should ask someone else to review it, or just "not yet/right now"?

**Comment by weihanglo** [comment]
It is in my queue, and I am happy to review when ready that :)

**Pull request merged**
